main.go

```diff
@@ -111,15 +113,48 @@ func run(cfg langserver.Config) error {
 		}
 		defer lis.Close()
 
-		log.Println("langserver-go: listening on", *addr)
+		log.Println("langserver-go: listening for TCP connections on", *addr)
+
+		connectionCount := 0
+
 		for {
 			conn, err := lis.Accept()
 			if err != nil {
 				return err
 			}
-			jsonrpc2.NewConn(context.Background(), jsonrpc2.NewBufferedStream(conn, jsonrpc2.VSCodeObjectCodec{}), handler, connOpt...)
+			connectionCount = connectionCount + 1
+			connectionID := connectionCount
+			log.Printf("langserver-go: received incoming connection #%d\n", connectionID)
+			jsonrpc2Connection := jsonrpc2.NewConn(context.Background(), jsonrpc2.NewBufferedStream(conn, jsonrpc2.VSCodeObjectCodec{}), handler, connOpt...)
+			go func() {
+				<-jsonrpc2Connection.DisconnectNotify()
+				log.Printf("langserver-go: disconnected connection #%d\n", connectionID)
+			}()
 		}
 
+	case "websocket":
+		mux := http.NewServeMux()
+
+		connectionCount := 0
+
+		mux.HandleFunc("/", func(responseWriter http.ResponseWriter, request *http.Request) {
+			var upgrader = websocket.Upgrader{CheckOrigin: func(r *http.Request) bool { return true }}
+			connection, err := upgrader.Upgrade(responseWriter, request, nil)
+			if err != nil {
+				log.Println("error upgrading HTTP to WebSocket:", err)
+			}
+			defer connection.Close()
+			connectionCount = connectionCount + 1
+			connectionID := connectionCount
+			log.Printf("langserver-go: received incoming connection #%d\n", connectionID)
+			<-jsonrpc2.NewConn(context.Background(), wsjsonrpc2.NewObjectStream(connection), langserver.NewHandler(cfg), connOpt...).DisconnectNotify()
+			log.Printf("langserver-go: disconnected connection #%d\n", connectionID)
+		})
+
+		log.Println("langserver-go: listening for WebSocket connections on", *addr)
+		http.ListenAndServe(*addr, mux)
```

Comment:
if this returns an error we probably want to print it and return it. For example if it can't bind to the address, etc it will fail.

Reply:
I added a log statement for `http.ListenAndServe`